services/motion/builtin/builtin.go

```diff
@@ -452,3 +334,181 @@ func (ms *builtIn) PlanHistory(
 	defer ms.mu.RUnlock()
 	return ms.state.PlanHistory(req)
 }
+
+// DoCommand supports two commands which are specified through the command map
+//   - DoPlan generates and returns a Trajectory for a given motionpb.MoveRequest without executing it
+//     required key: DoPlan
+//     input value: a motionpb.MoveRequest which will be used to create a Trajectory
+//     output value: a motionplan.Trajectory specified as a map (the mapstructure.Decode function is useful for decoding this)
+//   - DoExecute takes a Trajectory and executes it
+//     required key: DoExecute
+//     input value: a motionplan.Trajectory
+//     output value: a bool
+func (ms *builtIn) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
+	ms.mu.RLock()
+	defer ms.mu.RUnlock()
+	operation.CancelOtherWithLabel(ctx, builtinOpLabel)
+
+	resp := make(map[string]interface{}, 0)
+	if req, ok := cmd[DoPlan]; ok {
+		bytes, err := json.Marshal(req)
+		if err != nil {
+			return nil, err
+		}
+		var moveReqProto pb.MoveRequest
+		err = json.Unmarshal(bytes, &moveReqProto)
+		if err != nil {
+			return nil, err
+		}
+		moveReq, err := motion.MoveReqFromProto(&moveReqProto)
+		if err != nil {
+			return nil, err
+		}
+		plan, err := ms.plan(ctx, moveReq)
+		if err != nil {
+			return nil, err
+		}
+		resp[DoPlan] = plan.Trajectory()
+	}
+	if req, ok := cmd[DoExecute]; ok {
+		var trajectory motionplan.Trajectory
+		if err := mapstructure.Decode(req, &trajectory); err != nil {
+			return nil, err
+		}
+		if err := ms.execute(ctx, trajectory); err != nil {
+			return nil, err
+		}
+		resp[DoExecute] = true
+	}
+	return resp, nil
+}
+
+func (ms *builtIn) plan(ctx context.Context, req motion.MoveReq) (motionplan.Plan, error) {
+	if req.Destination == nil {
```

Comment:
We should probably generally validate all contents of the MoveRequest

Reply:
We are already, the only components of the MoveReq that are not allowed to be nil are the Destination and the componentName which are already being checked